File: Power Query/PQ_Challenge_153.R
library(tidyverse)
library(readxl)

input = read_excel("Power Query/PQ_Challenge_153.xlsx", range = "A1:C13") %>% 
  janitor::clean_names()
test  = read_excel("Power Query/PQ_Challenge_153.xlsx", range = "E1:G5") %>%
  janitor::clean_names()

input$pilot = factor(input$pilot, levels = unique(input$pilot), ordered = TRUE)
test$pilot = factor(test$pilot, levels = unique(test$pilot), ordered = TRUE)

result = input %>%
  group_by(pilot) %>%
  mutate(prev_landing = lag(flight_end, default = NA_POSIXct_),
         flight_time = flight_end - flight_start,
         rest_time = flight_start - prev_landing) %>%
  summarise(fly_time = sum(flight_time, na.rm = TRUE),
            rest_time = sum(rest_time, na.rm = TRUE)) %>%
  mutate(fly_time = as.numeric(fly_time, units = "hours") %>% round(2),
         rest_time = as.numeric(rest_time, units = "hours") %>% round(2)) %>%
  arrange(pilot) %>%
  ungroup() %>%
  mutate(fly_time = ifelse(fly_time == 0, NA, fly_time),
         rest_time = ifelse(rest_time == 0, NA, rest_time))

identical(result, test)
# [1] TRUE

File: Power Query/100-199/153/PQ_Challenge_153.py
import pandas as pd

input_data = pd.read_excel("PQ_Challenge_153.xlsx", usecols="A:C", nrows=13)
input_data.columns = [c.strip().lower() for c in input_data.columns]
test = pd.read_excel("PQ_Challenge_153.xlsx", usecols="E:G", nrows=5)
test.columns = [c.strip().lower() for c in test.columns]

result = input_data.copy()
result["prev_landing"] = result.groupby("pilot")["flight_end"].shift()
result["flight_time"] = result["flight_end"] - result["flight_start"]
result["rest_time"] = result["flight_start"] - result["prev_landing"]
result = result.groupby("pilot", as_index=False).agg(fly_time=("flight_time", "sum"), rest_time=("rest_time", "sum"))
result["fly_time"] = result["fly_time"].dt.total_seconds().div(3600).round(2)
result["rest_time"] = result["rest_time"].dt.total_seconds().div(3600).round(2)
result["fly_time"] = result["fly_time"].where(result["fly_time"] != 0)
result["rest_time"] = result["rest_time"].where(result["rest_time"] != 0)

print(result.equals(test))
